packages/context/src/context-view.ts

```diff
@@ -0,0 +1,131 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import * as debugFactory from 'debug';
+import {promisify} from 'util';
+import {Binding} from './binding';
+import {BindingFilter} from './binding-filter';
+import {Context} from './context';
+import {
+  ContextObserver,
+  ContextEventType,
+  Subscription,
+} from './context-observer';
+import {Getter} from './inject';
+import {ResolutionSession} from './resolution-session';
+import {resolveList, ValueOrPromise} from './value-promise';
+const debug = debugFactory('loopback:context:view');
+const nextTick = promisify(process.nextTick);
+
+/**
+ * `ContextView` provides a view for a given context chain to maintain a live
+ * list of matching bindings and their resolved values within the context
+ * hierarchy.
+ *
+ * This class is the key utility to implement dynamic extensions for extension
+ * points. For example, the RestServer can react to `controller` bindings even
+ * they are added/removed/updated after the application starts.
+ *
+ */
+export class ContextView<T = unknown> implements ContextObserver {
+  protected _cachedBindings: Readonly<Binding<T>>[] | undefined;
+  protected _cachedValues: T[] | undefined;
```

Comment:
Would it be too strict to assume that the values are all in the same type?

Reply:
No. We can relax as follows:

```ts
ContextView<Type1 | Type2>
```